src/queue.rs

```diff
@@ -0,0 +1,148 @@
+const QUEUE_WORKER_COUNT: u8 = 2;
+
+fn schedule_based_on_failure_count(failure_count: i32) -> Option<OffsetDateTime> {
+    if failure_count >= MAX_RETRY {
+        None
+    } else {
+        Some(
+            OffsetDateTime::now_utc()
+                + Duration::from_millis(
+                    4000_u64.pow(failure_count.try_into().unwrap()) + fastrand::u64(0..15000),
+                ),
+        )
+    }
+}
+
+pub async fn one(db: &Db, job_state: &SharedJobState) -> Result<Option<Model>, DbErr> {
+    let tx = db.begin().await?;
+    let model = if let Some(model) = next(&tx).await? {
+        let mut active_model = model.into_active_model();
+        let mut job = active_model
+            .job
+            .take()
+            .expect("queue jobs should always have a Job");
+        let result = job.perform(job_state, &tx).await;
+        active_model.job = Set(job);
+        active_model.updated_at = Set(OffsetDateTime::now_utc());
+        match result {
+            Ok(Some(job)) => {
+                active_model.status = Set(JobStatus::Success);
+                active_model.scheduled_at = Set(None);
+                let mut next_job = ActiveModel::from(job);
+                next_job.parent_id = Set(Some(*active_model.id.as_ref()));
+                let next_job = next_job.insert(&tx).await?;
+                active_model.result = Set(Some(JobResult::Child(next_job.id)));
+            }
+
+            Ok(None) => {
+                active_model.scheduled_at = Set(None);
+                active_model.status = Set(JobStatus::Success);
+                active_model.result = Set(Some(JobResult::Complete));
+            }
+
+            Err(e) if e.is_retryable() => {
+                active_model.failure_count = Set(active_model.failure_count.as_ref() + 1);
+                let reschedule =
+                    schedule_based_on_failure_count(*active_model.failure_count.as_ref());
+                active_model.status =
+                    Set(reschedule.map_or(JobStatus::Failed, |_| JobStatus::Pending));
+                active_model.scheduled_at = Set(reschedule);
+                active_model.result = Set(Some(JobResult::Error(e)));
+            }
+
+            Err(e) => {
+                active_model.failure_count = Set(active_model.failure_count.as_ref() + 1);
+                active_model.scheduled_at = Set(None);
+                active_model.status = Set(JobStatus::Failed);
+                active_model.result = Set(Some(JobResult::Error(e)));
+            }
+        }
+        Some(active_model.update(&tx).await?)
+    } else {
+        None
+    };
+    tx.commit().await?;
+    Ok(model)
+}
+
+fn spawn(join_set: &mut JoinSet<()>, db: &Db, job_state: &Arc<SharedJobState>) {
+    let db = db.clone();
+    let job_state = Arc::clone(job_state);
+    join_set.spawn(async move {
+        loop {
+            match one(&db, &job_state).await {
+                Err(e) => {
+                    eprintln!("job error {e}");
```

Comment:
It'd be better to use `tracing::error` so that we get a structured message that can be ingested by our log indexer.

Reply:
I see now that `divviup-api` doesn't currently use `tracing` at all. I think that should be addressed for consistency of logging across the components of Divvi Up, but this PR isn't the place to do it: #69